README.md
[![GoDoc reference](https://img.shields.io/badge/godoc-reference-blue.svg)](https://pkg.go.dev/code.posterity.life/srp/v2)

```golang
go get code.posterity.life/srp/v2@latest
```

# Secure Remote Password

Package srp is a Go implementation of
[Secure Remote Password](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol)
protocol as defined by [RFC 2945](https://tools.ietf.org/html/rfc2945) and
[RFC 5054](https://tools.ietf.org/html/rfc5054).

> SRP is an authentication method that allows the use
> of user names and passwords over unencrypted channels without
> revealing the password to an eavesdropper. SRP also supplies a
> shared secret at the end of the authentication sequence that can be
> used to generate encryption keys.

SRP is used by leading privacy-conscious companies such as
[Apple](https://support.apple.com/guide/security/escrow-security-for-icloud-keychain-sec3e341e75d/web),
[1Password](https://blog.1password.com/developers-how-we-use-srp-and-you-can-too/),
[ProtonMail](https://protonmail.com/blog/cryptographic-architecture-response/),
and yours truly.

## Protocol

Conceptually, SRP is not different from how most of us think about
authentication; the client signs up by storing a _secret_ on the server, and to
login, it must prove to that server that it knows it.

With SRP, the client first registers by storing a cryptographic value (`verifier`)
derived from its password on the server. To login, they both exchange a
series of opaque values but never the user's password or the `verifier`. Trust
can be established at the end of the process because for the server,
only the client who knows the `verifier` could have sent those values,
and vice versa.

SRP comes with four major benefits:

1. For the end-user, the familiar experience of using a username and a password
   remains fundamentally the same;
2. Server cannot leak a password it never saw;
3. After registration, both client and server can formally verify each
   other's identities without needing a third-party (e.g. CA);
4. Sessions can be secured with an extra layer of encryption on top of TLS.

### Params selection

SRP requires the client and the server to agree on a given set of parameters,
namely a Diffie-Hellman (DH) group, a hash function, and a key derivation
function.

All the DH groups defined in [RFC 5054](https://tools.ietf.org/html/rfc5054)
are available. You can use any hash function you would like
(e.g. `SHA256`, [Blake2b](https://pkg.go.dev/golang.org/x/crypto/blake2b)), and
the same goes for key derivation
(e.g. [Argon2](https://pkg.go.dev/golang.org/x/crypto/argon2),
[Scrypt](https://pkg.go.dev/golang.org/x/crypto/scrypt) or
[PBKDF2](https://pkg.go.dev/golang.org/x/crypto/pbkdf2)).

The example below shows the DH group 16 used in conjunction with `SHA256` and
[Argon2](https://pkg.go.dev/golang.org/x/crypto/argon2):

```golang
import (
  "runtime"
  "github.com/posterity/srp"
  "golang.org/x/crypto/argon2"

  _ "crypto/sha256"
)

// KDFArgon2 uses Argon2.
func KDFArgon2(username, password string, salt []byte) ([]byte, error) {
  p := []byte(username + ":" + password)
  key := argon2.IDKey(p, salt, 3, 256 * 1048576, runtime.NumCPU(), 32)
  return key, nil
}

// Params instance using DH group 16, SHA256 for hashing and Argon2 as a KDF.
var params = &srp.Params{
  Name: "DH16–SHA256–Argon2",
  Group: srp.RFC5054Group4096,
  Hash: crypto.SHA256,
  KDF: KDFArgon2,
}
```

### User Registration

During user registration, the client must send the server a `verifier`; a
value safely derived from the user's password with a unique random salt.

```go
tp, err := srp.ComputeVerifier(params, username, password, srp.NewSalt())
if err != nil {
  log.Fatalf("error computing verifier: %v", err)
}

// The verifier can be accessed as tp.Verifier().

// On the server, it's recommended to store the verifier along with
// the username and the salt used to compute it, so sending the whole
// triplet tp ([]byte) is more appropriate.
Send(tp)
```

The `Triplet` returned by `ComputeVerifier` encapsulates three variables into a
single byte array that the server can store:

- Username
- Verifier
- Salt

It's important for the server to treat the triplet with care, as it contains
a secret value (`verifier`) which should never be shared with anyone.

The `salt` value it contains however should be made available publicly to
_anyone who asks_ via a public URL.

### Login

When it's time to authenticate a user, client and server follow a three-step
process:

1. `client` and `server` exchange ephemeral public keys `A` and `B`,
   respectively;
2. `client` computes a proof and sends it to the server;
3. `server` checks the client's proof and sends the client a proof of their own.

#### Client-side

On the client side, the first step is to initialize a `Client`.

```go
var (
  username  = "alice@example.com"
  password  = "p@$$w0rd"
  salt      []byte // Retrieved from the server
)
client, err := srp.NewClient(params, username, password, salt)
if err != nil {
  log.Fatal(err)
}
```

All the values must match those used to create the verifier that was stored
on the server. The `salt` should be retrievable from the server without
requiring prior authentication.

The next step is to send the ephemeral public key `A` to the server:

```go
A := client.A()

// Send A to the server
```

The server will do the same, sending their ephemeral public key `B` instead.
Configure it on the client as following:

```go
var B []byte // Received from the server

client.SetB(B)
```

Next, compute the client proof and send it to the server.

```go
M1, err := client.ComputeM1()
if err != nil {
  log.Fatalf("error computing proof: %v", err)
}

// send M1 to the server
```

If the server accepts the client's proof, they will send their own server proof.

```go
var M2 []byte // Received from the server

ok, err := client.CheckM2(M2)
if err != nil {
  log.Fatalf("error checking M2: %v", err)
}
if !ok {
  log.Fatalf("server is not authentic")
}
```

At this stage, the client and the server can trust each other, and can
(optionally) use a shared encryption key to secure their session from this
point on.

```go
sharedKey, err := client.SessionKey()
if err != nil {
  log.Fatalf("error computing key: %v", err)
}

// sharedKey is a 256 bit key which was computed
// locally.
```

#### Server-side

The process on the server-side is very similar to the above, with one key
difference: the server must first receive and verify the client's proof (`M1`)
before it computes and shares its own (`M2`).

```go
var (
  triplet srp.Triplet                             // Retrieved from the server
)
server, err := srp.NewServer(params, username, password, salt)
if err != nil {
  log.Fatal(err)
}
```

The next step is to wait for the user to send their ephemeral public key `A`
to configure it on the server.

```go
var A []byte // received from the client

if err := server.setA(A); err != nil {
  log.Fatal("error configuring A: %v", err)
}
```

If no error is caught, the next step is to send to server's ephemeral public
key `B` to the client.

```go
B := server.B()

// send B to the client
```

Now the server must wait for the client to submit their proof `M1`.

```go
var M1 []byte   // Received from the client

ok, err := server.CheckM1(M1)
if err != nil {
  log.Fatalf("error verifying M1: %v", err)
}

if !ok {
  log.Fatalf("client is not authentic")
}
```

If this verification fails, the process must stop at this point, and no further
information should be shared with the client over this session. A new `Server`
instance will need to be created and the negotiation restarted.

If successful, the server can consider the client as _authentic_, but it
still needs to send its own proof `M2`.

```go
M2, err := server.ComputeM2()
if err != nil {
  log.Fatalf("error computing M2: %v", err)
}

// send M2 to the client
```

If the client accepts the proof, they can both consider each other as
_authentic_ and compute their shared session key to encrypt their exchanges
and protect themselves from eavesdropping.

```go
sharedKey, err := server.SessionKey()
if err != nil {
  log.Fatalf("error computing key: %v", err)
}

// sharedKey is a 256 bit key which was computed
// locally.
```

## Implementation

SRP is protocol-agnostic and can be implemented on top of any existing
client/server architecture.

At Posterity, we use a custom websocket protocol, but a simple HTTP API would
be equally suitable. In any case, the process can usually be completed in
two round-trips, excluding the request needed to retrieve the `salt` value
of the user:

```plain
(Client) 👧🏼  ---------→ A
                        B   ←--------- 👨🏽 (Server)

(Client) 👧🏼  ---------→ M1
                        M2  ←--------- 👨🏽 (Server)

```

If you're using a stateless architecture (e.g. REST), the state of a `Server`
can be saved and restored using `Server.Save` and `RestoreServer` respectively.
Bear in mind that a `Server`'s internal state contains the user's `verifier`,
and should therefore be handled appropriately.

A secure connection between the client and the server is a necessity,
especially when the client first needs to send their `verifier` to the server.

## Session Encryption

SRP defines a way for the client and the server to independently compute a
strong but ephemeral encryption key which they can use to secure their
communications during a session.

At Posterity, we use
[Encrypted-Content-Encoding for HTTP](https://github.com/posterity/ece) to set
that in motion, using the shared key to encrypt all client-server exchanges
with AES-256-GCM after login.

## Contributions

Contributions are welcome via Pull Requests.

## About us

What if you're hit by a bus tomorrow? [Posterity](https://posterity.life) helps
you make a plan in the event something happens to you.


CHANGELOG.md
# Changelog

## v2.0.1

- Requires Go 1.20;
- Now using Go's new subtle.XORBytes function;
- Added more tests.

## v1.1.1

- Usernames and passwords are now normalized to NFKD unicode strings;
- Added changelog.


srp_test.go
package srp

import (
	"bytes"
	"crypto"
	_ "crypto/sha1"
	_ "embed"
	"encoding/hex"
	"log"
	"testing"
)

// Test vectors imported from RFC 5054 – Appendix B
// https://datatracker.ietf.org/doc/html/rfc5054#appendix-B
var (
	params = &Params{
		Group: RFC5054Group1024,
		Hash:  crypto.SHA1,
		KDF:   RFC5054KDF,
	}
	I    = []byte("alice")
	P    = []byte("password123")
	x    = mustParseHex("94B7555A ABE9127C C58CCF49 93DB6CF8 4D16C124")
	salt = mustParseHex("BEB25379 D1A8581E B5A72767 3A2441EE")
	k    = mustParseHex("7556AA04 5AEF2CDD 07ABAF0F 665C3E81 8913186F")
	v    = mustParseHex(
		"7E273DE8 696FFC4F 4E337D05 B4B375BE B0DDE156 9E8FA00A 9886D812",
		"9BADA1F1 822223CA 1A605B53 0E379BA4 729FDC59 F105B478 7E5186F5",
		"C671085A 1447B52A 48CF1970 B4FB6F84 00BBF4CE BFBB1681 52E08AB5",
		"EA53D15C 1AFF87B2 B9DA6E04 E058AD51 CC72BFC9 033B564E 26480D78",
		"E955A5E2 9E7AB245 DB2BE315 E2099AFB",
	)
	a = mustParseHex(
		"60975527 035CF2AD 1989806F 0407210B C81EDC04 E2762A56 AFD529DD",
		"DA2D4393",
	)
	A = mustParseHex(
		"61D5E490 F6F1B795 47B0704C 436F523D D0E560F0 C64115BB 72557EC4",
		"4352E890 3211C046 92272D8B 2D1A5358 A2CF1B6E 0BFCF99F 921530EC",
		"8E393561 79EAE45E 42BA92AE ACED8251 71E1E8B9 AF6D9C03 E1327F44",
		"BE087EF0 6530E69F 66615261 EEF54073 CA11CF58 58F0EDFD FE15EFEA",
		"B349EF5D 76988A36 72FAC47B 0769447B",
	)
	b = mustParseHex(
		"E487CB59 D31AC550 471E81F0 0F6928E0 1DDA08E9 74A004F4 9E61F5D1",
		"05284D20",
	)
	B = mustParseHex(
		"BD0C6151 2C692C0C B6D041FA 01BB152D 4916A1E7 7AF46AE1 05393011",
		"BAF38964 DC46A067 0DD125B9 5A981652 236F99D9 B681CBF8 7837EC99",
		"6C6DA044 53728610 D0C6DDB5 8B318885 D7D82C7F 8DEB75CE 7BD4FBAA",
		"37089E6F 9C6059F3 88838E7A 00030B33 1EB76840 910440B1 B27AAEAE",
		"EB4012B7 D7665238 A8E3FB00 4B117B58",
	)
	u = mustParseHex("CE38B959 3487DA98 554ED47D 70A7AE5F 462EF019")
	S = mustParseHex(
		"B0DC82BA BCF30674 AE450C02 87745E79 90A3381F 63B387AA F271A10D",
		"233861E3 59B48220 F7C4693C 9AE12B0A 6F67809F 0876E2D0 13800D6C",
		"41BB59B6 D5979B5C 00A172B4 A2A5903A 0BDCAF8A 709585EB 2AFAFA8F",
		"3499B200 210DCC1F 10EB3394 3CD67FC8 8A2F39A4 BE5BEC4E C0A3212D",
		"C346D7E4 74B29EDE 8A469FFE CA686E5A",
	)
	K = params.Hash.New().Sum(S.Bytes())[:params.Hash.New().Size()]
)

// assertEqualBytes fails t if wanted != got
func assertEqualBytes(t *testing.T, name string, wanted, got []byte) {
	t.Helper()

	if !bytes.Equal(got, wanted) {
		t.Fatalf("%s - bytes don't match", name)
	}
}

// assertNotNil fails t if got == nil
func assertNotNil(t *testing.T, name string, got []byte) {
	t.Helper()

	if got == nil {
		t.Fatalf("%s should not be nil", name)
	}
}

func TestServerKeyPair(t *testing.T) {
	b, B := newServerKeyPair(params, k, v)
	if b == bigZero {
		t.Fatal("b should not be bigZero")
	}

	if !isValidEphemeralKey(params, B) {
		t.Fatal("B is an invalid ephemeral key")
	}
}

func TestClientKeyPair(t *testing.T) {
	a, A := newClientKeyPair(params)
	if a == bigZero {
		t.Fatal("a should not be bigZero")
	}

	if !isValidEphemeralKey(params, A) {
		t.Fatal("A is an invalid ephemeral key")
	}
}

func TestComputeLittleU(t *testing.T) {
	got, err := computeLittleU(params, A, B)
	if err != nil {
		t.Fatal(err)
	}

	assertEqualBytes(t, "u", u.Bytes(), got.Bytes())
}

func TestComputeLittleUInvalidA(t *testing.T) {
	_, err := computeLittleU(params, nil, B)
	if err == nil {
		t.Fatal("expected err")
	}
}

func TestComputeLittleK(t *testing.T) {
	got, err := computeLittleK(params)
	if err != nil {
		t.Fatal(err)
	}

	assertEqualBytes(t, "k", k.Bytes(), got.Bytes())
}

func TestComputeS(t *testing.T) {
	t.Run("Server", func(t *testing.T) {
		got, err := computeServerS(params, v, u, A, b)
		if err != nil {
			t.Fatal(err)
		}

		assertEqualBytes(t, "S", S.Bytes(), got.Bytes())
	})

	t.Run("Client", func(t *testing.T) {
		got, err := computeClientS(params, k, x, u, B, a)
		if err != nil {
			t.Fatal(err)
		}
		t.Log(hex.EncodeToString(got.Bytes()))
		assertEqualBytes(t, "S", S.Bytes(), got.Bytes())
	})
}

func TestComputeLittleX(t *testing.T) {
	got, err := params.KDF(string(I), string(P), salt.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	assertEqualBytes(t, "x", x.Bytes(), got)
}

func TestComputeVerifier(t *testing.T) {
	got, err := ComputeVerifier(params, string(I), string(P), salt.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	assertEqualBytes(t, "v", v.Bytes(), got.Verifier())
}

func TestComputeM(t *testing.T) {
	M1, err := computeM1(params, I, salt.Bytes(), A, B, K)
	if err != nil {
		t.Fatal(err)
	}

	if _, err := computeM2(params, A, M1, K); err != nil {
		t.Fatal(err)
	}
}

func TestNewServer(t *testing.T) {
	s, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}
	assertEqualBytes(t, "v", s.triplet.Verifier(), v.Bytes())
	if s.b == nil || s.b.Cmp(bigZero) == 0 {
		t.Fatal("s.b is nil or invalid")
	}
	if s.xB == nil || !isValidEphemeralKey(s.params, s.xB) {
		t.Fatal("s.xB is nil or invalid")
	}
	assertEqualBytes(t, "username", []byte(s.triplet.Username()), I)
	assertEqualBytes(t, "salt", s.triplet.Salt(), salt.Bytes())
}

func TestServerSetA(t *testing.T) {
	s, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	if err := s.SetA(A.Bytes()); err != nil {
		t.Fatal(err)
	}
	assertEqualBytes(t, "A", s.xA.Bytes(), A.Bytes())

	assertNotNil(t, "s.m1", s.m1.Bytes())
	assertNotNil(t, "s.m2", s.m2.Bytes())
	assertNotNil(t, "s.xS", s.xS.Bytes())
	assertNotNil(t, "s.xK", s.xK)
}

func TestNewClient(t *testing.T) {
	c, err := NewClient(params, string(I), string(P), salt.Bytes())
	if err != nil {
		t.Fatal(err)
	}
	assertEqualBytes(t, "x", c.x.Bytes(), x.Bytes())
	if c.a == nil || c.a.Cmp(bigZero) == 0 {
		t.Fatal("c.a is nil or invalid")
	}
	if c.xA == nil || !isValidEphemeralKey(c.params, c.xA) {
		t.Fatal("c.xA is nil or invalid")
	}
	assertEqualBytes(t, "username", c.username, I)
	assertEqualBytes(t, "salt", c.salt, salt.Bytes())
}

func TestClientSetB(t *testing.T) {
	c, err := NewClient(params, string(I), string(P), salt.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	if err := c.SetB(B.Bytes()); err != nil {
		t.Fatal(err)
	}
	assertEqualBytes(t, "B", c.xB.Bytes(), B.Bytes())
	if c.m1 == nil {
		t.Fatal("c.m1 should not be nil")
	}
	if c.m2 == nil {
		t.Fatal("c.m1 should not be nil")
	}
	if c.xS == nil {
		t.Fatal("c.xS should not be nil")
	}
	if c.xK == nil {
		t.Fatal("c.xK should not be nil")
	}

	assertNotNil(t, "c.m1", c.m1.Bytes())
	assertNotNil(t, "c.m2", c.m2.Bytes())
	assertNotNil(t, "c.xS", c.xS.Bytes())
	assertNotNil(t, "c.xK", c.xK)
}

func TestCheckM1(t *testing.T) {
	M1, err := computeM1(params, I, salt.Bytes(), A, B, K)
	if err != nil {
		t.Fatal(err)
	}

	if !checkProof(M1.Bytes(), M1.Bytes()) {
		t.Fatal("checkProof failed")
	}

	if checkProof(M1.Bytes(), bigZero.Bytes()) {
		t.Fatal("checkProof should have failed")
	}
}

func TestSession(t *testing.T) {
	client, err := NewClient(params, string(I), string(P), salt.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	server, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	if err := server.SetA(client.A()); err != nil {
		t.Fatal(err)
	}

	if err := client.SetB(server.B()); err != nil {
		t.Fatal(err)
	}

	M1, err := client.ComputeM1()
	if err != nil {
		t.Fatal(err)
	}

	// Server Check M1
	clientOK, err := server.CheckM1(M1)
	if err != nil {
		t.Fatal(err)
	}
	if !clientOK {
		t.Fatal("client is not authentic")
	}

	M2, err := server.ComputeM2()
	if err != nil {
		t.Fatal(err)
	}

	// Client checks M2
	serverOK, err := client.CheckM2(M2)
	if err != nil {
		t.Fatal(err)
	}
	if !serverOK {
		t.Fatal("server is not authentic")
	}

	// Compare keys
	cK, err := client.SessionKey()
	if err != nil {
		t.Fatal(err)
	}
	sK, err := server.SessionKey()
	if err != nil {
		t.Fatal(err)
	}
	assertEqualBytes(t, "K", cK, sK)
}

func TestSessionProofOrder(t *testing.T) {
	server, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	if err := server.SetA(A.Bytes()); err != nil {
		t.Fatal(err)
	}

	if _, err := server.ComputeM2(); err == nil {
		t.Fatal("expected an error here, because server.CheckM1 has not been called yet")
	}
}

func TestNewSalt(t *testing.T) {
	b := NewSalt()
	if len(b) != SaltLength {
		t.Fatal("invalid length")
	}
}

// Send is a noop used for examples.
func Send(any) {}

// Receive is a noop used for examples.
func Receive() []byte {
	return nil
}

// SendEncrypted is a nop used for examples.
func SendEncrypted(key, data []byte) {}

// Query is a nop used for examples.
func Query(any) Triplet { return nil }

// Example of a client session.
func ExampleClient() {
	var (
		username = "alice@example.com"
		password = "some-password"
	)

	// Request the user's salt from the server.
	// The server should send it to whoever asks.
	salt := Receive()

	// Create a client, specifying the same params used on the server.
	client, err := NewClient(params, username, password, salt)
	if err != nil {
		log.Fatal(err)
	}

	// Send A to the server.
	A := client.A()
	Send(A)

	// Receive B from the server,
	// then configure it on the client.
	B := Receive()
	if err := client.SetB(B); err != nil {
		log.Fatalf("invalid B received from the server: %v", err)
	}

	// Compute the proof M1,
	// then send it to the server.
	M1, err := client.ComputeM1()
	if err != nil {
		log.Fatalf("failed to compute M1: %v", err)
	}
	Send(M1)

	// If the server accepts the client's proof (M1), it will
	// send a proof of their own (M2).
	M2 := Receive()
	valid, err := client.CheckM2(M2)
	if err != nil {
		log.Fatalf("failed to verify server proof M2: %v", err)
	}
	if valid == false {
		log.Fatalf("server is not authentic: %v", err)
	}

	// At this stage, the client and the server
	// have proved to each other that they know
	// the same secret.
	//
	// They can both consider each other as authentic
	// and legitimate.

	// They also share a common key they both derived independently
	// from the process
	K, err := client.SessionKey()
	if err != nil {
		log.Fatalf("failed to access shared session key: %v", err)
	}

	// K can optionally be used to encrypt/decrypt all exchanges between
	// them moving forward.
	SendEncrypted(K, []byte("hello, world!"))
}

// Example of a server session.
func ExampleServer() {
	// Load the user's Triplet from the persistent
	// storage.
	var user Triplet = Query("alice@example.com")

	// Create a server, specifying the same params used on the client.
	server, err := NewServer(params, user.Username(), user.Salt(), user.Verifier())
	if err != nil {
		log.Fatal(err)
	}

	// Send B to the client.
	B := server.B()
	Send(B)

	// Receive A from the client,
	// then configure it on the server.
	A := Receive()
	if err := server.SetA(A); err != nil {
		log.Fatalf("invalid A received from the client: %v", err)
	}

	// The server needs to verify the client's proof first,
	// so it must wait to receive M1.
	M1 := Receive()

	// Verify the client proof M1.
	// The process must be interrupted if valid is false, or
	// an error occurred.
	valid, err := server.CheckM1(M1)
	if err != nil {
		log.Fatalf("failed to verify client proof M1: %v", err)
	}
	if valid == false {
		log.Fatalf("client is not authentic: %v", err)
	}

	// The client proved they're authentic, so it's safe
	// to compute the server proof (M2) and send it over.
	M2, err := server.ComputeM2()
	if err != nil {
		log.Fatalf("failed to compute M1: %v", err)
	}
	Send(M2)

	// At this stage, the server should consider
	// the client as authentic and requests from it
	// should be fulfilled.

	// They both share a common key they both derived
	// independently from the process,
	K, err := server.SessionKey()
	if err != nil {
		log.Fatalf("failed to access session key: %v", err)
	}

	// K can optionally be used to encrypt/decrypt all exchanges between
	// them moving forward.
	SendEncrypted(K, []byte("hello, world!"))
}

// The verifier is calculated on the client, and sent to the
// server for storage along with the username and salt used to
// compute it as a triplet.
func ExampleComputeVerifier() {
	const (
		username = "bob@example.com"
		password = "p@$$w0rd"
	)
	tp, err := ComputeVerifier(params, username, password, NewSalt())
	if err != nil {
		log.Fatalf("failed to compute verifier: %v", err)
	}

	// The verifier can be accessed via the returned triplet tp
	// as tp.Verifier().

	// On the server, it's recommended to store the verifier along with
	// the username and the salt used to compute it, so sending the whole
	// triplet ([]byte) is more appropriate.
	Send(tp)
}

// This example shows how to create a Params instance that ensures strict
// compatibility with [RFC5054] (Not recommended in production).
//
// [RFC5054]: https://datatracker.ietf.org/doc/html/rfc5054
func ExampleParams() {
	params := &Params{
		Group: RFC5054Group1024,
		Hash:  crypto.SHA1,
		KDF:   RFC5054KDF,
	}

	_, err := NewClient(params, "username", "p@$$w0rd", NewSalt())
	if err != nil {
		log.Fatal(err)
	}
}


groups_test.go
package srp

import (
	"crypto"
	"testing"
)

func TestClone(t *testing.T) {
	var (
		name = "g"
		h    = crypto.SHA256
		kdf  = func(username, password string, salt []byte) ([]byte, error) {
			return []byte("test"), nil
		}
	)
	g := RFC5054Group2048.Clone(name, h, kdf)
	if g.Name != name {
		t.Error("failed to set name")
	}
	if g.Hash.String() != h.String() {
		t.Error("failed to set hash")
	}
	if b, err := g.KDF("", "", []byte("")); string(b) != "test" && err != nil {
		t.Error("failed to set KDF")
	}
}


server_test.go
package srp

import "testing"

func TestRestoreServerJSON(t *testing.T) {
	server, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}

	if err := server.SetA(A.Bytes()); err != nil {
		t.Fatal(err)
	}

	state, err := server.Save()
	if err != nil {
		t.Fatal(err)
	}

	restored, err := RestoreServer(params, state)
	if err != nil {
		t.Fatal(err)
	}

	assertEqualBytes(t, "triplet", server.triplet, restored.triplet)
	assertEqualBytes(t, "b", server.b.Bytes(), restored.b.Bytes())
	assertEqualBytes(t, "B", server.xB.Bytes(), restored.xB.Bytes())
	assertEqualBytes(t, "A", server.xA.Bytes(), restored.xA.Bytes())
	assertEqualBytes(t, "S", server.xS.Bytes(), restored.xS.Bytes())
	assertEqualBytes(t, "K", server.xK, restored.xK)
}

func TestServerReset(t *testing.T) {
	s, err := NewServer(params, string(I), salt.Bytes(), v.Bytes())
	if err != nil {
		t.Fatal(err)
	}
	s.SetA(A.Bytes())

	M1, err := computeM1(params, I, salt.Bytes(), A, s.xB, s.xK)
	if err != nil {
		t.Fatal(err)
	}
	if ok, err := s.CheckM1(M1.Bytes()); !ok {
		t.Fatalf("M1 not verified: %v", err)
	}

	if !s.verifiedM1 {
		t.Fatal("expected M1 to be verified")
	}

	s.Reset(params, string(I), salt.Bytes(), v.Bytes())
	if _, err := s.CheckM1(M1.Bytes()); err != ErrServerNoReady {
		t.Fatal("expected server to not be ready")
	}
	if s.verifiedM1 {
		t.Fatal("expected M1 to not be verified")
	}
}


params.go
package srp

import (
	"crypto"
	"fmt"

	"errors"
	"math/big"
	"strings"

	_ "crypto/sha1" //#nosec
	_ "embed"       // Embedding RFC5054 DH groups
)

var (
	//go:embed groups/1024.txt
	hex1024 string

	//go:embed groups/1536.txt
	hex1536 string

	//go:embed groups/2048.txt
	hex2048 string

	//go:embed groups/3072.txt
	hex3072 string

	//go:embed groups/4096.txt
	hex4096 string

	//go:embed groups/6144.txt
	hex6144 string

	//go:embed groups/8192.txt
	hex8192 string
)

// KDF is the signature of a key derivation function.
type KDF func(username, password string, salt []byte) ([]byte, error)

// MustParseHex returns a *big.Int instance
// from the given hex string, or panics.
func mustParseHex(parts ...string) *big.Int {
	builder := new(strings.Builder)
	for _, p := range parts {
		p = strings.TrimSpace(p)
		p = strings.ReplaceAll(p, " ", "")
		p = strings.ReplaceAll(p, "\r", "")
		p = strings.ReplaceAll(p, "\n", "")
		builder.WriteString(p)
	}

	n, ok := new(big.Int).SetString(builder.String(), 16)
	if !ok {
		panic(errors.New("failed to load params N"))
	}

	return n
}

// RFC5054KDF is the [KDF] defined in [RFC5054].
//
// 	x = SHA(s | SHA(U | ":" | p))
//
// Deprecated: This KDF function is only provided for compatibility
// with [RFC5054] and for testing purposes. It is not recommended
// for production use. Instead, use a key derivation function
// [KDF] designed for password hashing such as [Argon2],
// [Scrypt] or [PBKDF2].
//
// [RFC5054]: https://datatracker.ietf.org/doc/html/rfc5054
// [Argon2]: https://pkg.go.dev/golang.org/x/crypto/argon2
// [Scrypt]: https://pkg.go.dev/golang.org/x/crypto/scrypt
// [PBKDF2]: https://pkg.go.dev/golang.org/x/crypto/pbkdf2
func RFC5054KDF(username, password string, salt []byte) ([]byte, error) {
	h := crypto.SHA1.New()
	h.Write([]byte(fmt.Sprintf("%s:%s", username, password)))
	digest := h.Sum(nil)[:h.Size()]

	h.Reset()
	h.Write(salt)
	h.Write(digest)
	digest = h.Sum(nil)[:h.Size()]
	return digest, nil
}

// Params represents the DH group, the hash and
// key derivation function that a client and server
// jointly agreed to use.
//
//   import (
//     "runtime"
//     "github.com/posterity/srp"
//   	 "golang.org/x/crypto/argon2"
//
//   	 _ "crypto/sha256"
// 	 )
//
// 	 func KDFArgon2(username, password string, salt []byte) ([]byte, error) {
//   	 p := []byte(username + ":" + password)
//   	 key := argon2.IDKey(p, salt, 3, 256 * 1048576, runtime.NumCPU(), 32)
//   	 return key, nil
// 	 }
//
// 	 var params = &srp.Params{
//   	 Name: "DH16–SHA256–Argon2",
//   	 Group: srp.RFC5054Group4096,
//   	 Hash: crypto.SHA256,
//   	 KDF: KDFArgon2,
// 	 }
type Params struct {
	Name  string
	Group *Group
	Hash  crypto.Hash
	KDF   KDF
}

// hashBytes returns the hash of a.
func (p *Params) hashBytes(a []byte) []byte {
	h := p.Hash.New()
	h.Write(a)
	return h.Sum(nil)[:h.Size()]
}

// String returns the name of p.
func (p *Params) String() string {
	return p.Name
}

// Group represents a Diffie-Hellman group.
type Group struct {
	ID           string
	Generator    *big.Int
	N            *big.Int
	ExponentSize int
}

// Diffie-Hellman group 2.
//
// Deprecated: This group is not recommended
// for production-use.
var RFC5054Group1024 = &Group{
	ID:           "2",
	Generator:    big.NewInt(2),
	N:            mustParseHex(hex1024),
	ExponentSize: 32,
}

// Diffie-Hellman group 5.
//
// Deprecated: This group is not recommended
// for production-use.
var RFC5054Group1536 = &Group{
	ID:           "5",
	Generator:    big.NewInt(2),
	N:            mustParseHex(hex1536),
	ExponentSize: 23,
}

// Diffie-Hellman group 14, 15, 16, 17 and 18
// defined in [RFC5054].
//
// [RFC5054]: https://datatracker.ietf.org/doc/html/rfc5054
var (
	RFC5054Group2048 = &Group{
		ID:           "14",
		Generator:    big.NewInt(2),
		N:            mustParseHex(hex2048),
		ExponentSize: 27,
	}

	RFC5054Group3072 = &Group{
		ID:           "15",
		Generator:    big.NewInt(5),
		N:            mustParseHex(hex3072),
		ExponentSize: 32,
	}

	RFC5054Group4096 = &Group{
		ID:           "16",
		Generator:    big.NewInt(5),
		N:            mustParseHex(hex4096),
		ExponentSize: 38,
	}

	RFC5054Group6144 = &Group{
		ID:           "17",
		Generator:    big.NewInt(5),
		N:            mustParseHex(hex6144),
		ExponentSize: 43,
	}

	RFC5054Group8192 = &Group{
		ID:           "18",
		Generator:    big.NewInt(19),
		N:            mustParseHex(hex8192),
		ExponentSize: 48,
	}
)
